### DISCUSSION TITLE
[Feature Request] official guide for wiring TextMate grammars

### DISCUSSION THREAD

**Discussion by souporserious** [microsoft/monaco-editor#3830] [Q&A]
### Context

- [X] This issue is not a bug report. *(please use a different template for reporting a bug)*
- [X] This issue is not a duplicate of an existing issue. *(please use the [search](https://github.com/microsoft/monaco-editor/issues) to find existing issues)*

### Description

Hello, I've been struggling to find a solution that works with TextMate grammars and the latest releases for `monaco-editor` and `monaco-editor-webpack-plugin`. I've tried following the linked `monaco-tm` package and the issue here https://github.com/microsoft/monaco-editor/issues/1915, but it seems some of the theme helpers `generateTokensCSSForColorMap`, `TokenizationRegistry`, and `Color` are no longer available. I even read the [wonderful blog post](https://code.visualstudio.com/blogs/2017/02/08/syntax-highlighting-optimizations) about the deep dive into how it's all working to try and better diagnose the issue and just can't seem to get any solution to work.

I previously had this working using `monaco-editor-textmate`, but that has recently broke between `monaco-editor` versions `0.29.1` and `0.30.0` https://github.com/zikaari/monaco-editor-textmate/issues/26.

I've managed to distill the wiring of textmate down to the solution below based on `monaco-editor-textmate` and [theia](https://github.com/eclipse-theia/theia/). I know it's not the most performant route per se, but I was hoping to use monaco as an enhancement on top of server-rendered shiki code blocks so this will be loaded on demand if someone wants to play around with code examples.

```tsx
import type { Registry, StateStack } from 'vscode-textmate'
import { INITIAL } from 'vscode-textmate'
import * as monaco from 'monaco-editor'

/** Wires up monaco-editor with monaco-textmate */
export function wireTextMateGrammars(
  /** TmGrammar `Registry` this wiring should rely on to provide the grammars. */
  registry: Registry,

  /** `Map` of language ids (string) to TM names (string). */
  languages: Map<string, string>,

  /** The monaco editor instance to wire up. */
  editor: monaco.editor.ICodeEditor
) {
  const tokenTheme = editor['_themeService'].getColorTheme().tokenTheme
  const defaultForeground = tokenTheme._root._mainRule._foreground

  return Promise.all(
    Array.from(languages.keys()).map(async (languageId) => {
      const grammar = await registry.loadGrammar(languages.get(languageId))

      monaco.languages.setTokensProvider(languageId, {
        getInitialState: () => new TokenizerState(INITIAL),
        tokenize: (line: string, state: TokenizerState) => {
          const result = grammar.tokenizeLine(line, state.ruleStack)

          return {
            endState: new TokenizerState(result.ruleStack),
            tokens: result.tokens.map((token) => {
              const scopes = token.scopes.slice(0)

              for (let i = scopes.length - 1; i >= 0; i--) {
                const scope = scopes[i]
                const foreground = tokenTheme._match(scope)._foreground

                if (foreground !== defaultForeground) {
                  return {
                    ...token,
                    scopes: scope,
                  }
                }
              }

              return {
                ...token,
                scopes: scopes[scopes.length - 1],
              }
            }),
          }
        },
      })
    })
  )
}

class TokenizerState implements monaco.languages.IState {
  constructor(private _ruleStack: StateStack) {}

  public get ruleStack(): StateStack {
    return this._ruleStack
  }

  public clone(): TokenizerState {
    return new TokenizerState(this._ruleStack)
  }

  public equals(other: monaco.languages.IState): boolean {
    return (
      other instanceof TokenizerState &&
      (other === this || other.ruleStack === this.ruleStack)
    )
  }
}
```

For some reason between those versions of the editor, the matched foreground (`tokenTheme._match(scope)._foreground`) has changed and is now returning a different foreground in some cases causing the syntax highlighting to be off.

I was going to file as a bug, but since there is no official guide and these are accessing private APIs it makes sense they are prone to breaking. Is it possible this is something on the roadmap or in the meantime there is a solution to workaround this until official guides/utilities might be available? I appreciate any help and understand there are a lot of moving parts here to make this all work.

**Comment by souporserious** [comment]
@alexdima sorry to ping, but do you have any advice here? I'm really stuck on this issue 😕 my latest attempt was forking the monaco-tm example and following your answer [here](https://github.com/microsoft/monaco-editor/issues/2239#issuecomment-752242294), but I just get black text I think possibly because the colorMap is only returning `[null, "#000000", "#FFFFFF"]`.

For context, I'm trying to create a [Next.js plugin](https://github.com/souporserious/next-monaco/blob/monaco-tm-fork/next-monaco/editor/index.tsx) that sets up TextMate grammars automatically.

**Comment by tamayika** [comment]
I don't know this is suitable for you, but you can [monaco-vscode-api](https://github.com/CodinGame/monaco-vscode-api) to use textmate grammer for syntax highlighting.
This currently works on monaco-editor 0.34.x and work in progress for 0.36.x at [here](https://github.com/CodinGame/monaco-vscode-api/pull/75))

**Reply by souporserious** [reply]
That library looks suitable from what I can tell, thank you for the link! Looks like they just merged that PR 🎉 I'm going to play around with it and see if I can get something working.

**Comment by caleb1248** [comment]
I also struggled with the same problem, and I ended up making [this thing](https://github.com/caleb1248/monaco-vscode-textmate). It works pretty well and I use it for most of my projects involving `monaco-editor`.
